src/main/java/com/fasterxml/jackson/core/io/NumberInput.java

```diff
@@ -190,14 +190,55 @@ public static long parseLong19(char[] ch, int off, boolean negative)
      */
     public static long parseLong(String s)
     {
-        // Ok, now; as the very first thing, let's just optimize case of "fake longs";
-        // that is, if we know they must be ints, call int parsing
-        int length = s.length();
-        if (length <= 9) {
-            return (long) parseInt(s);
+        /* Ok: let's keep strategy simple: ignoring optional minus sign,
+         * we'll accept 1 - 19 digits and parse things efficiently;
+         * otherwise just defer to JDK parse functionality.
+         */
+        char c = s.charAt(0);
+        int len = s.length();
+        boolean neg = (c == '-');
+        int offset = 1;
+        // must have 1 - 19 digits after optional sign:
+        // negative?
+        if (neg) {
+            if (len == 1 || len > 20) {
+                return Long.parseLong(s);
+            }
+            c = s.charAt(offset++);
+        } else {
+            if (len > 19) {
+                return Long.parseLong(s);
+            }
+        }
+        if (c > '9' || c < '0') {
+            return Integer.parseInt(s);
+        }
+        long num = c - '0';
+        if (offset < len) {
+            c = s.charAt(offset++);
+            if (c > '9' || c < '0') {
+                return Long.parseLong(s);
+            }
+            num = (num * 10) + (c - '0');
+            if (offset < len) {
+                c = s.charAt(offset++);
+                if (c > '9' || c < '0') {
+                    return Long.parseLong(s);
+                }
+                num = (num * 10) + (c - '0');
+                // Let's just loop if we have more than 3 digits:
+                if (offset < len) {
+                    do {
+                        c = s.charAt(offset++);
+                        if (c > '9' || c < '0') {
+                            return Long.parseLong(s);
+                        }
+                        num = (num * 10) + (c - '0');
```

Comment:
I haven't understood all the callers. How do you make sure you don't overflow on something like 9223372036854775808?

Reply:
this overflows with the new code so this is a bug - I've created https://github.com/FasterXML/jackson-core/pull/879 to extend test coverage

If you'd like to create a PR with the improved code you added to the benchmark project, that would be great - I'll close this PR.